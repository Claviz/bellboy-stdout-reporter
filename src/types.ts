export interface IStream {
    receivedRows: number;
    destinations: { [id: number]: IStreamDestination };
    bytes: number;
    streamId: number;
    streamInfo?: any[];
    finished: boolean;
    speed: number;
}

export interface IStreamDestination {
    rowsGenerated: number,
    rowsGenerationFails: number,
    batchesLoaded: number,
    batchLoadFails: number,
    batchesTransformed: number,
    batchTransformFails: number,
    bytesLoaded: 0;
}

export interface LogItem {
    header: string;
    type?: 'success' | 'fail' | 'wait';
    lines?: { [key: string]: string };
    children?: LogItem[];
}

export interface IReporterConfig {
    interval?: number;
}