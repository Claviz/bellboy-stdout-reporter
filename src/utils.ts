const stringifyObject = require('stringify-object');

export function getCurrentTimestamp() {
    return new Date().toISOString();
}

export function stringify(obj: any) {
    const maxArraySize = 3;
    if (Array.isArray(obj) && obj.length > maxArraySize) {
        const limitText = `â€¦ [+${obj.length - maxArraySize} items]`
        obj = obj.slice(0, maxArraySize);
        obj.push(limitText);
    }
    let pretty = stringifyObject(obj, {
        indent: '  ',
        singleQuotes: false,
        inlineCharacterLimit: 200,
    });
    return pretty;
}