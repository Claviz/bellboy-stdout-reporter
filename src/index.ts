import { Job, Reporter } from 'bellboy';
import chalk from 'chalk';
import filesize from 'filesize';
import prettyMs from 'pretty-ms';

import { IReporterConfig, IStream, LogItem } from './types';
import { getCurrentTimestamp, stringify } from './utils';

let globalJobId = 0;

export default class StdoutReporter extends Reporter {

    private interval: number;
    private verbose: boolean;

    constructor(config?: IReporterConfig) {
        super();
        this.interval = 5000;
        this.verbose = false;
        if (config) {
            this.interval = config.interval || 5000;
            this.verbose = config.verbose || false;
        }
    }

    report(job: Job) {
        let jobId = globalJobId.toString();
        if ((job as any).jobName) {
            jobId += ` (${(job as any).jobName})`;
        }
        globalJobId++;
        let stream: IStream;
        const startedMs = (new Date()).getTime();

        let timeoutObj: NodeJS.Timeout | null;

        function logItem(item: LogItem, indent: boolean = true, withTimestamp: boolean = true) {
            if (indent) {
                console.log();
            }
            let header = item.header;
            if (item.type === 'success') {
                header = `${chalk.green('\u221A')} ${header}`;
            } else if (item.type === 'fail') {
                header = `${chalk.red('x')} ${header}`;
            } else if (item.type === 'wait') {
                header = `${chalk.yellow('⌛︎')} ${header}`;
            }
            console.log(chalk.bold(header));
            if (withTimestamp) {
                item.lines = {
                    'Timestamp': getCurrentTimestamp(),
                    ...item.lines,
                }
            }
            if (item.lines || item.children) {
                console.group();
                if (item.lines) {
                    const lineKeys = Object.keys(item.lines);
                    for (let i = 0; i < lineKeys.length; i++) {
                        const key = lineKeys[i];
                        const value = item.lines[key];
                        console.log(`${chalk.bold(key)}: ${value}`);
                    }
                }
                if (item.children) {
                    for (let i = 0; i < item.children.length; i++) {
                        logItem(item.children[i], false, false);
                    }
                }
                console.groupEnd();
            }
        }

        function getStreamStatus() {
            let header = `Job #${jobId}. Stream #${stream.streamId} `;
            header += stream.streamInfo && stream.streamInfo.length ? `(${stream.streamInfo.join(', ')}) ` : '';
            if (stream.finished) {
                header += 'processed';
            } else {
                const speed = stream.speed ? ` (${Math.round(stream.speed)} rows/sec)` : '';
                header += `processing${speed}…`;
            }
            const streamLogItem: LogItem = {
                type: stream.finished ? 'success' : 'wait',
                header,
                lines: {
                    'Rows': chalk.green(`${stream.receivedRows} received (${filesize(stream.bytes)})`)
                },
                children: [],
            };
            const destinationIds = Object.keys(stream.destinations);
            for (let i = 0; i < destinationIds.length; i++) {
                const destinationId = destinationIds[i];
                const destination = stream.destinations[Number(destinationId)];
                const destinationLogItem: LogItem = {
                    header: `Destination #${destinationId}`,
                    lines: {},
                }
                if (destination.rowsGenerated || destination.rowsGenerationFails) {
                    const arr = [];
                    if (destination.rowsGenerated) {
                        arr.push(chalk.green(`${destination.rowsGenerated} generated`));
                    }
                    if (destination.rowsGenerationFails) {
                        arr.push(chalk.red(`${destination.rowsGenerationFails} failed`));
                    }
                    destinationLogItem.lines!['Row generation'] = arr.join(', ');
                }
                if (destination.batchesTransformed || destination.batchTransformFails) {
                    const arr = [];
                    if (destination.batchesTransformed) {
                        arr.push(chalk.green(`${destination.batchesTransformed} transformed`));
                    }
                    if (destination.batchTransformFails) {
                        arr.push(chalk.red(`${destination.batchTransformFails} failed`));
                    }
                    destinationLogItem.lines!['Row transformation'] = arr.join(', ');
                }
                if (destination.batchesLoaded || destination.batchLoadFails) {
                    const arr = [];
                    if (destination.batchesLoaded) {
                        arr.push(chalk.green(`${destination.batchesLoaded} loaded(${filesize(destination.bytesLoaded)})`));
                    }
                    if (destination.batchLoadFails) {
                        arr.push(chalk.red(`${destination.batchLoadFails} failed`));
                    }
                    destinationLogItem.lines!['Batch load'] = arr.join(', ');
                }
                streamLogItem.children!.push(destinationLogItem);
            }
            logItem(streamLogItem);
        }

        function getDestination(destinationIndex: number) {
            let destination = stream.destinations[destinationIndex];
            if (!destination) {
                destination = {
                    batchLoadFails: 0,
                    batchesLoaded: 0,
                    rowsGenerated: 0,
                    rowsGenerationFails: 0,
                    batchTransformFails: 0,
                    batchesTransformed: 0,
                    bytesLoaded: 0,
                };
                stream.destinations[destinationIndex] = destination;
            }
            return destination;
        }

        job.on('startProcessingRow', async (data) => {
            stream.receivedRows++;
            stream.bytes += Buffer.byteLength(JSON.stringify(data), 'utf8');
            if (this.verbose) {
                logItem({
                    type: 'success',
                    header: `Job #${jobId}. Stream #${stream.streamId}. Row received`,
                    lines: {
                        'Data': stringify(data),
                    }
                });
            }
        });
        job.on('rowGenerated', async (destinationIndex, data) => {
            const destination = getDestination(destinationIndex);
            destination.rowsGenerated++;
            if (this.verbose) {
                logItem({
                    type: 'success',
                    header: `Job #${jobId}. Stream #${stream.streamId}. Destination #${destinationIndex} Row generated`,
                    lines: {
                        'Data': stringify(data),
                    }
                });
            }
        });
        job.on('rowGenerationError', async (destinationIndex, data, error) => {
            const destination = getDestination(destinationIndex);
            destination.rowsGenerationFails++;
            logItem({
                type: 'fail',
                header: `Job #${jobId}. Stream #${stream.streamId}. Destination #${destinationIndex}. Row generation failed`,
                lines: {
                    'Source row': stringify(data),
                    'Error': error,
                }
            });
        });
        job.on('transformingBatchError', async (destinationIndex, data, error) => {
            const destination = getDestination(destinationIndex);
            destination.batchTransformFails++;
            logItem({
                type: 'fail',
                header: `Job #${jobId}. Stream #${stream.streamId}. Destination #${destinationIndex}. Batch transform failed`,
                lines: {
                    'Source data': stringify(data),
                    'Error': error,
                }
            });
        });
        job.on('transformedBatch', async (destinationIndex, data) => {
            const destination = getDestination(destinationIndex);
            destination.batchesTransformed++;
            if (this.verbose) {
                logItem({
                    type: 'success',
                    header: `Job #${jobId}. Stream #${stream.streamId}. Destination #${destinationIndex} Batch transformed`,
                    lines: {
                        'Data': stringify(data),
                    }
                });
            }
        });
        job.on('loadingBatchError', async (destinationIndex, data, error) => {
            const destination = getDestination(destinationIndex);
            destination.batchLoadFails++;
            logItem({
                type: 'fail',
                header: `Job #${jobId}. Stream #${stream.streamId}. Destination #${destinationIndex}. Batch load failed`,
                lines: {
                    'Source data': stringify(data),
                    'Error': error,
                }
            });
        });
        job.on('loadedBatch', async (destinationIndex, data) => {
            const destination = getDestination(destinationIndex);
            destination.batchesLoaded++;
            destination.bytesLoaded += Buffer.byteLength(JSON.stringify(data), 'utf8');
            if (this.verbose) {
                logItem({
                    type: 'success',
                    header: `Job #${jobId}. Stream #${stream.streamId}. Destination #${destinationIndex} Batch loaded`,
                    lines: {
                        'Data': stringify(data),
                    }
                });
            }
        });
        job.on('startProcessing', async () => {
            logItem({
                type: 'success',
                header: `Job #${jobId} started`,
            });
        });
        job.on('endProcessing', async () => {
            const finishedMs = (new Date()).getTime();
            const durationMs = finishedMs - startedMs;
            logItem({
                type: 'success',
                header: `Job #${jobId} completed`,
                lines: {
                    'Job duration': prettyMs(durationMs),
                }
            });
        });
        job.on('processingError', async (err) => {
            logItem({
                type: 'fail',
                header: `Job #${jobId} failed`,
                lines: {
                    'Error': err,
                }
            });
        });
        let timePassed = 0;
        job.on('startProcessingStream', async (...args) => {
            stream = {
                receivedRows: 0,
                destinations: [],
                bytes: 0,
                streamId: stream && stream.streamId ? stream.streamId : 0,
                streamInfo: args,
                finished: false,
                speed: 0,
            };
            timeoutObj = setInterval(() => {
                if (!stream.finished) {
                    timePassed += this.interval;
                    const speed = stream.receivedRows / (timePassed / 1000);
                    stream.speed = speed;
                    getStreamStatus();
                }
            }, this.interval);
        });
        job.on('endProcessingStream', async (...argumets) => {
            if (timeoutObj) {
                clearInterval(timeoutObj);
                timeoutObj = null;
            }
            stream.finished = true;
            getStreamStatus();
            stream.streamId++;
        });
    }
}