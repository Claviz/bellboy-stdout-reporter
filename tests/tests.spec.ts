import { DynamicProcessor, Job } from 'bellboy';
import stripAnsi from 'strip-ansi';

import { StdoutReporter } from '../src';
import { DummyDestination, FaultyDestination, timeout, FaultyProcessor, WithInfoProcessor, MultipleStreamProcessor } from './helpers';

jest.mock('pretty-ms', () => (() => 'pretty-ms-mock'));
jest.mock('../src/utils', () => ({
    ...jest.requireActual('../src/utils'),
    getCurrentTimestamp: () => 'timestamp-mock'
}));

let consoleData: string[] = [];

beforeAll(async () => {
});

beforeEach(async () => {
    consoleData = [];
    let indent = 0;
    console.log = function (message: string) {
        message = message ? message : '';
        consoleData.push('    '.repeat(indent) + stripAnsi(message));
    }
    console.group = function () {
        indent++;
    }
    console.groupEnd = function () {
        indent--;
    }
});

afterEach(async () => {
});

afterAll(async () => {
});

it('logs successfull bellboy job', async () => {
    const destination = new DummyDestination();
    const processor = new DynamicProcessor({
        generator: async function* () {
            yield `test`;
        },
    });
    const job = new Job(processor, [destination], {
        reporters: [new StdoutReporter()],
    });
    await job.run();
    expect(consoleData).toMatchSnapshot();
});

it('logs load exceptions', async () => {
    const destination = new FaultyDestination();
    const processor = new DynamicProcessor({
        generator: async function* () {
            yield `test`;
        },
    });
    const job = new Job(processor, [destination], {
        reporters: [new StdoutReporter()],
    });
    await job.run();
    expect(consoleData).toMatchSnapshot();
});

it('logs intermediate results after reaching some interval', async () => {
    const destination = new DummyDestination();
    const processor = new DynamicProcessor({
        generator: async function* () {
            yield `test`;
            await timeout(1000);
        },
    });
    const job = new Job(processor, [destination], {
        reporters: [new StdoutReporter({ interval: 1000 })],
    });
    await job.run();
    expect(consoleData).toMatchSnapshot();
});

it('logs failing bellboy job', async () => {
    const destination = new DummyDestination();
    const processor = new FaultyProcessor();
    const job = new Job(processor, [destination], {
        reporters: [new StdoutReporter()],
    });
    await job.run();
    expect(consoleData).toMatchSnapshot();
});

it('logs destination batch transformations', async () => {
    const destination = new DummyDestination({
        batchTransformer: async function (rows) {
            if (rows[0] === `shouldSuccess`) {
                return rows;
            } else {
                throw new Error(`Oh, snap!`);
            }
        },
        batchSize: 1,
    });
    const processor = new DynamicProcessor({
        generator: async function* () {
            yield `shouldSuccess`;
            yield `shouldFail`;
        },
    });
    const job = new Job(processor, [destination], {
        reporters: [new StdoutReporter()],
    });
    await job.run();
    expect(consoleData).toMatchSnapshot();
});

it('logs destination record generations', async () => {
    const destination = new DummyDestination({
        recordGenerator: async function* (row) {
            if (row === `shouldSuccess`) {
                yield row;
            } else {
                throw new Error(`Oh, snap!`);
            }
        },
    });
    const processor = new DynamicProcessor({
        generator: async function* () {
            yield `shouldSuccess`;
            yield `shouldFail`;
        },
    });
    const job = new Job(processor, [destination], {
        reporters: [new StdoutReporter()],
    });
    await job.run();
    expect(consoleData).toMatchSnapshot();
});

it('logs stream info if stream emits one', async () => {
    const destination = new DummyDestination();
    const processor = new WithInfoProcessor();
    const job = new Job(processor, [destination], {
        reporters: [new StdoutReporter()],
    });
    await job.run();
    expect(consoleData).toMatchSnapshot();
});

it('logs multiple streams', async () => {
    const destination = new DummyDestination();
    const processor = new MultipleStreamProcessor();
    const job = new Job(processor, [destination], {
        reporters: [new StdoutReporter()],
    });
    await job.run();
    expect(consoleData).toMatchSnapshot();
});