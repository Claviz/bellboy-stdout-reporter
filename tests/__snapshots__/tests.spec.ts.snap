// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logs destination batch transformations 1`] = `
Array [
  "",
  "√ Job #4 started",
  "    Timestamp: timestamp-mock",
  "",
  "x Job #4. Stream #0. Destination #0. Batch transform failed",
  "    Timestamp: timestamp-mock",
  "    Source data: [\\"shouldFail\\"]",
  "    Error: Error: Oh, snap!",
  "",
  "√ Job #4. Stream #0 processed",
  "    Timestamp: timestamp-mock",
  "    Rows: 2 received (27 B)",
  "    Destination #0",
  "        Row transformation: 1 transformed, 1 failed",
  "        Batch load: 2 loaded(31 B)",
  "",
  "√ Job #4 completed",
  "    Timestamp: timestamp-mock",
  "    Job duration: pretty-ms-mock",
]
`;

exports[`logs destination record generations 1`] = `
Array [
  "",
  "√ Job #5 started",
  "    Timestamp: timestamp-mock",
  "",
  "x Job #5. Stream #0. Destination #0. Row generation failed",
  "    Timestamp: timestamp-mock",
  "    Source row: \\"shouldFail\\"",
  "    Error: Error: Oh, snap!",
  "",
  "√ Job #5. Stream #0 processed",
  "    Timestamp: timestamp-mock",
  "    Rows: 2 received (27 B)",
  "    Destination #0",
  "        Row generation: 1 generated, 1 failed",
  "        Batch load: 1 loaded(17 B)",
  "",
  "√ Job #5 completed",
  "    Timestamp: timestamp-mock",
  "    Job duration: pretty-ms-mock",
]
`;

exports[`logs failing bellboy job 1`] = `
Array [
  "",
  "√ Job #3 started",
  "    Timestamp: timestamp-mock",
  "",
  "x Job #3 failed",
  "    Timestamp: timestamp-mock",
  "    Job duration: pretty-ms-mock",
  "    Error: Error: Oh, snap!",
]
`;

exports[`logs intermediate results after reaching some interval 1`] = `
Array [
  "",
  "√ Job #2 started",
  "    Timestamp: timestamp-mock",
  "",
  "⌛︎ Job #2. Stream #0 processing (1 rows/sec)…",
  "    Timestamp: timestamp-mock",
  "    Rows: 1 received (6 B)",
  "",
  "√ Job #2. Stream #0 processed",
  "    Timestamp: timestamp-mock",
  "    Rows: 1 received (6 B)",
  "    Destination #0",
  "        Batch load: 1 loaded(8 B)",
  "",
  "√ Job #2 completed",
  "    Timestamp: timestamp-mock",
  "    Job duration: pretty-ms-mock",
]
`;

exports[`logs load exceptions 1`] = `
Array [
  "",
  "√ Job #1 started",
  "    Timestamp: timestamp-mock",
  "",
  "x Job #1. Stream #0. Destination #0. Batch load failed",
  "    Timestamp: timestamp-mock",
  "    Source data: [\\"test\\"]",
  "    Error: Error: Oh, snap!",
  "",
  "√ Job #1. Stream #0 processed",
  "    Timestamp: timestamp-mock",
  "    Rows: 1 received (6 B)",
  "    Destination #0",
  "        Batch load: 1 failed",
  "",
  "√ Job #1 completed",
  "    Timestamp: timestamp-mock",
  "    Job duration: pretty-ms-mock",
]
`;

exports[`logs multiple streams 1`] = `
Array [
  "",
  "√ Job #7 started",
  "    Timestamp: timestamp-mock",
  "",
  "√ Job #7. Stream #0 processed",
  "    Timestamp: timestamp-mock",
  "    Rows: 1 received (6 B)",
  "    Destination #0",
  "        Batch load: 1 loaded(8 B)",
  "",
  "√ Job #7. Stream #1 processed",
  "    Timestamp: timestamp-mock",
  "    Rows: 0 received (0 B)",
  "",
  "√ Job #7 completed",
  "    Timestamp: timestamp-mock",
  "    Job duration: pretty-ms-mock",
]
`;

exports[`logs stream info if stream emits one 1`] = `
Array [
  "",
  "√ Job #6 started",
  "    Timestamp: timestamp-mock",
  "",
  "√ Job #6. Stream #0 (some, info) processed",
  "    Timestamp: timestamp-mock",
  "    Rows: 1 received (6 B)",
  "    Destination #0",
  "        Batch load: 1 loaded(8 B)",
  "",
  "√ Job #6 completed",
  "    Timestamp: timestamp-mock",
  "    Job duration: pretty-ms-mock",
]
`;

exports[`logs successfull bellboy job 1`] = `
Array [
  "",
  "√ Job #0 started",
  "    Timestamp: timestamp-mock",
  "",
  "√ Job #0. Stream #0 processed",
  "    Timestamp: timestamp-mock",
  "    Rows: 1 received (6 B)",
  "    Destination #0",
  "        Batch load: 1 loaded(8 B)",
  "",
  "√ Job #0 completed",
  "    Timestamp: timestamp-mock",
  "    Job duration: pretty-ms-mock",
]
`;

exports[`logs successfull bellboy job in verbose mode 1`] = `
Array [
  "",
  "√ Job #8 started",
  "    Timestamp: timestamp-mock",
  "",
  "√ Job #8. Stream #0. Row received",
  "    Timestamp: timestamp-mock",
  "    Data: \\"test\\"",
  "",
  "√ Job #8. Stream #0. Destination #0 Row generated",
  "    Timestamp: timestamp-mock",
  "    Data: \\"test\\"",
  "",
  "√ Job #8. Stream #0. Destination #0 Batch transformed",
  "    Timestamp: timestamp-mock",
  "    Data: [\\"test\\"]",
  "",
  "√ Job #8. Stream #0. Destination #0 Batch loaded",
  "    Timestamp: timestamp-mock",
  "    Data: [\\"test\\"]",
  "",
  "√ Job #8. Stream #0 processed",
  "    Timestamp: timestamp-mock",
  "    Rows: 1 received (6 B)",
  "    Destination #0",
  "        Row generation: 1 generated",
  "        Row transformation: 1 transformed",
  "        Batch load: 1 loaded(8 B)",
  "",
  "√ Job #8 completed",
  "    Timestamp: timestamp-mock",
  "    Job duration: pretty-ms-mock",
]
`;

exports[`logs truncated array 1`] = `
Array [
  "",
  "√ Job #9 started",
  "    Timestamp: timestamp-mock",
  "",
  "√ Job #9. Stream #0. Row received",
  "    Timestamp: timestamp-mock",
  "    Data: \\"test\\"",
  "",
  "√ Job #9. Stream #0. Row received",
  "    Timestamp: timestamp-mock",
  "    Data: \\"test\\"",
  "",
  "√ Job #9. Stream #0. Row received",
  "    Timestamp: timestamp-mock",
  "    Data: \\"test\\"",
  "",
  "√ Job #9. Stream #0. Row received",
  "    Timestamp: timestamp-mock",
  "    Data: \\"test\\"",
  "",
  "√ Job #9. Stream #0. Row received",
  "    Timestamp: timestamp-mock",
  "    Data: \\"test\\"",
  "",
  "√ Job #9. Stream #0. Row received",
  "    Timestamp: timestamp-mock",
  "    Data: \\"test\\"",
  "",
  "√ Job #9. Stream #0. Row received",
  "    Timestamp: timestamp-mock",
  "    Data: \\"test\\"",
  "",
  "√ Job #9. Stream #0. Row received",
  "    Timestamp: timestamp-mock",
  "    Data: \\"test\\"",
  "",
  "√ Job #9. Stream #0. Row received",
  "    Timestamp: timestamp-mock",
  "    Data: \\"test\\"",
  "",
  "√ Job #9. Stream #0. Row received",
  "    Timestamp: timestamp-mock",
  "    Data: \\"test\\"",
  "",
  "√ Job #9. Stream #0. Destination #0 Batch loaded",
  "    Timestamp: timestamp-mock",
  "    Data: [\\"test\\", \\"test\\", \\"test\\", \\"… [+7 items]\\"]",
  "",
  "√ Job #9. Stream #0 processed",
  "    Timestamp: timestamp-mock",
  "    Rows: 10 received (60 B)",
  "    Destination #0",
  "        Batch load: 1 loaded(71 B)",
  "",
  "√ Job #9 completed",
  "    Timestamp: timestamp-mock",
  "    Job duration: pretty-ms-mock",
]
`;
